Intersect.Server/Entities/Combat/Dash.cs

```diff
@@ -50,40 +51,63 @@ public partial class Dash
             bool zdimensionPass = false
         )
         {
-            var n = 0;
             en.MoveTimer = 0;
             Range = 0;
             for (var i = 1; i <= range; i++)
             {
-                n = en.CanMove(Direction);
-                if (n == -5) //Check for out of bounds
+                if (!en.CanMoveInDirection(Direction, out var blockerType, out var entityType))
                 {
-                    return;
-                } //Check for blocks
-
-                if (n == -2 && blockPass == false)
-                {
-                    return;
-                } //Check for ZDimensionTiles
-
-                if (n == -3 && zdimensionPass == false)
-                {
-                    return;
-                } //Check for active resources
-
-                if (n == (int) EntityType.Resource && activeResourcePass == false)
-                {
-                    return;
-                } //Check for dead resources
-
-                if (n == (int) EntityType.Resource && deadResourcePass == false)
-                {
-                    return;
-                } //Check for players and solid events
-
-                if (n == (int) EntityType.Player || n == (int) EntityType.Event)
-                {
-                    return;
+                    switch (blockerType)
+                    {
+                        case MovementBlockerType.OutOfBounds:
+                            return;
+
+                        case MovementBlockerType.MapAttribute:
+                            if (blockPass)
+                            {
+                                break;
+                            }
+
+                            return;
+
+                        case MovementBlockerType.ZDimension:
+                            if (zdimensionPass)
+                            {
+                                break;
+                            }
+
+                            return;
+
+                        case MovementBlockerType.Entity:
+                            switch (entityType)
+                            {
+                                case EntityType.Resource:
+                                    if (activeResourcePass || deadResourcePass)
+                                    {
+                                        break;
+                                    }
+
+                                    return;
```

Comment:
This does _not_ match the original logic.

Reply:
uhm, but this is functionally equivalent to this:

```csharp
                if (n == (int) EntityType.Resource && activeResourcePass == false)
                {
                    return;
                } //Check for dead resources

                if (n == (int) EntityType.Resource && deadResourcePass == false)
                {
                    return;
                }
```

The original code uses a series of conditional statements to check if the entity type is a resource and if certain flags are set, and then returns early if the conditions are not met.

The new one uses a switch statement to check the entity type, and then checks the same flags using a single if statement with a logical OR operator. If the conditions are met, the code block breaks out of the switch statement, otherwise it returns early.

So, both implementations achieve the same result: if the entity type is a resource and the necessary flags are not set, the code exits the block (return) and does not execute any further instructions.

Reply:
It is not.

Before
- if the blocking entity was a resource and
  - `activeResourcePass` was false and `deadResourcePass` was true, it would return
  - `deadResourcePass` was false and `activeResourcePass` was true, it would return
  
Your new code `break`s if either one is `true`, instead of `return`ing if either one is `false`. They are not functionally equivalent.

Reply:
oh 🤔

would something like this do it?

```csharp
switch (blockerType)
{
    case EntityType.Resource:
        if (!activeResourcePass && deadResourcePass)
        {
            return;
        }
        break;
    // other cases omitted for brevity
}
```